City-of-Helsinki/hel-fi-drupal-grants
# Feature/AU-1412: Custom Webform config ignore and config import by application type ID

# [AU-1412](https://helsinkisolutionoffice.atlassian.net/browse/AU-1412)

## What was done
This pull request implements new features to the `WebformImportCommands.php` file.

The new features are the following:

1. Added the possibility to add Webform configurations (and their translations) to a custom ignore list. This can be done by adding a `config_import_ignore` array containing the forms application type ID to the `/conf/cmi/grants_metadata.settings.yml` file in the following manner:

**Example (add to the end of the file):**

<img width="1722" alt="image" src="https://github.com/City-of-Helsinki/hel-fi-drupal-grants/assets/117807015/b6d90804-fdd2-4b19-a58b-1c39f17462cd">

```
config_import_ignore:
  - 29
  - 48
  - 49
  - 51
  - 47
  - 59
  - 66
  - 56
  - 50
  - 53
  - 65
```

2. Added the possibility to import a single Webforms configuration. This can be done by passing in the forms **application type ID** as a parameter when running the `drush gwi` command.

**Example:**
```
drush gwi 48
drush gwi 51
```

3. Added the possibility to bypass any ignored Webform configurations with a `force` flag. This flag makes it possible to import either **all** or **a single** Webforms configuration even if they are set to be ignored in the `config_import_ignore` array.

**Example:**
```
drush gwi --force
drush gwi 48 --force
```

## How to install

Make sure your instance is up and running on the correct branch.
  * `git checkout feature/AU-1412-webform-config-production-ignore`
  * `make fresh`
  * `make drush-cr`

## How to test

1. Run `drush gwi.` Check that everything is running normally.
2. Run `drush gwi 48` and `drush gwi 51`. Check that only these Webforms get imported.
3. Add a few Webforms to the config ignore list by adding the following to the `config_import_ignore` array inside the `/conf/cmi/grants_metadata.settings.yml` file.

```
config_import_ignore:
  - 48
  - 51
```
4. Run `drush gwi`. Check that the ignored forms are not imported.
5. Run `drush gwi --force`. Check that the ignored forms are imported.
6. Run `drush gwi 48`. Check that the form is not imported.
7. Run `drush gwi 48 --force`. Check that the form is imported.

## ToDo
The new Webform "versioning" that we are implementing might require code changes to this implementation. Further investigation needed when said features are in place.

[AU-1412]: https://helsinkisolutionoffice.atlassian.net/browse/AU-1412?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
